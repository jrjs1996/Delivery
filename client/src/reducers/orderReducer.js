import { FETCH_ORDERS, ADD_ORDER, COMPLETE_ORDER } from '../actions/types';

const initialState = {
  items: [],
  item: {},
};

export default function (state = initialState, action) {
  switch (action.type) {
    case FETCH_ORDERS:
      return {
        ...state,
        items: action.payload,
      };
    case ADD_ORDER:
      return {
        ...state,
        items: [action.payload, ...state.items],
      };
    case COMPLETE_ORDER:
      return {
        ...state,
        items: state.items.map((item) => {
          if (item._id === action.payload._id) return Object.assign(item, action.payload);
          return item;
        }),
      };
    default:
      return state;
  }
}
