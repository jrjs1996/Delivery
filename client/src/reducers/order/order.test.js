import orderReducer from './order';
import {
  ADD_ORDER,
  FETCH_ORDERS,
  UPDATE_ORDER,
  ADD_CUSTOMER,
  ADD_TO_CURRENT_ORDER,
  REMOVE_FROM_CURRENT_ORDER,
} from '../../actions/types';
import { ordersMock, createAction, menuItemsMock } from '../../tests/mocks';

const createState = (items, currentOrder) => ({
  items,
  currentOrder,
});

describe('Order reducer', () => {
  it('Initial state is empty', () => {
    const initialState = orderReducer(undefined, 'FAKE_ACTION');
    expect(initialState).toMatchSnapshot();
  });

  it("Returns given state when an action doesn't match", () => {
    const state = orderReducer({ mockState: 'MockState' }, 'FAKE_ACTION');
    expect(state).toMatchSnapshot();
  });

  it('Adds given order when calling ADD_ORDER', () => {
    const initialItems = { [ordersMock[0]._id]: ordersMock[0] };
    const state = createState(initialItems, undefined);
    const action = createAction(ADD_ORDER, ordersMock[1]);
    expect(orderReducer(state, action)).toMatchSnapshot();
  });

  it(`Adds item to current order with a count of 1 when currentOrder
      is empty and calling ADD_TO_CURRENT_ORDER`, () => {
    const action = createAction(ADD_TO_CURRENT_ORDER, menuItemsMock[0]);
    expect(orderReducer(undefined, action)).toMatchSnapshot();
  });

  it(`Adds to the items count when calling ADD_TO_CURRENT_ORDER and the
      item is already in the order`, () => {
    const initialCurrentOrder = {
      [menuItemsMock[0]._id]: {
        item: menuItemsMock[0],
        count: 1,
      },
    };
    const state = createState(undefined, initialCurrentOrder);
    const action = createAction(ADD_TO_CURRENT_ORDER, menuItemsMock[0]);
    expect(orderReducer(state, action)).toMatchSnapshot();
  });

  it('Populates orders when calling FETCH_ORDERS', () => {
    const action = createAction(FETCH_ORDERS, ordersMock);
    expect(orderReducer(createState(), action)).toMatchSnapshot();
  });

  it(`REMOVE_FROM_CURRENT_ORDER item count when
      removing an item from order and count is > 1`, () => {
    const initialCurrentOrder = {
      [menuItemsMock[0]._id]: {
        item: menuItemsMock[0],
        count: 2,
      },
    };
    const state = createState(undefined, initialCurrentOrder);
    const action = createAction(REMOVE_FROM_CURRENT_ORDER, menuItemsMock[0]._id);
    expect(orderReducer(state, action)).toMatchSnapshot();
  });

  it(`REMOVE_FROM_CURRENT_ORDER removes item when
      removing an item from order and count is <= 1`, () => {
    const initialCurrentOrder = {
      [menuItemsMock[0]._id]: {
        item: menuItemsMock[0],
        count: 1,
      },
    };
    const state = createState(undefined, initialCurrentOrder);
    const action = createAction(REMOVE_FROM_CURRENT_ORDER, menuItemsMock[0]._id);
    expect(orderReducer(state, action)).toMatchSnapshot();
  });

  it('Updates given order in items to be payload when calling UPDATE_ORDER', () => {
    const initialItems = {
      [ordersMock[0]._id]: ordersMock[0],
      [ordersMock[1]._id]: ordersMock[1],
    };
    const state = createState(initialItems, undefined);
    const action = createAction(UPDATE_ORDER, {
      _id: ordersMock[0]._id,
      address: 'UPDATED INFO',
    });
    expect(orderReducer(state, action)).toMatchSnapshot();
  });
});
