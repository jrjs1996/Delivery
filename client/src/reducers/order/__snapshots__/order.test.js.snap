// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Order reducer Adds given order when calling ADD_ORDER 1`] = `
Object {
  "currentOrder": undefined,
  "items": Object {
    "0": Object {
      "__v": 0,
      "_id": "0",
      "address": "232 Blue St.",
      "customerName": "Joe Green",
      "delivery": true,
      "items": Array [
        Object {
          "__v": 0,
          "_id": "0",
          "archived": false,
          "created": "Now",
          "description": "A hamburger (short: burger) is a sandwich consisting of one or more cooked
    patties of ground meat, usually beef, placed inside a sliced bread roll or bun. The
    patty may be pan fried, grilled, or flame broiled. Hamburgers are often served with
     cheese, lettuce, tomato, onion, pickles, bacon, or chiles; condiments such as ketchup,
     mayonnaise, mustard, relish, or 'special sauce'; and are frequently placed on sesame
    seed buns. A hamburger topped with cheese is called a cheeseburger.",
          "menuNumber": 1,
          "price": 10.5,
          "revision": 0,
          "title": "Hamburger",
        },
        Object {
          "__v": 0,
          "_id": "1",
          "archived": false,
          "created": "Now",
          "description": "A salad is a dish consisting of a mixture of small pieces of food, usually vegetables.
    [1][2] However, different varieties of salad may contain virtually any type of ready-to-eat food.
     Salads are typically served at room temperature or chilled, with notable exceptions such as south
     German potato salad which is served warm.",
          "menuNumber": 2,
          "price": 8.25,
          "revision": 0,
          "title": "Salad",
        },
      ],
      "orderCreated": "Mon Jul 01 2019 23:57:40 GMT-060",
      "stage": 0,
    },
    "1": Object {
      "__v": 0,
      "_id": "1",
      "address": "322 Park St.",
      "customerName": "John Anderson",
      "delivery": false,
      "items": Array [
        Object {
          "__v": 0,
          "_id": "1",
          "archived": false,
          "created": "Now",
          "description": "A salad is a dish consisting of a mixture of small pieces of food, usually vegetables.
    [1][2] However, different varieties of salad may contain virtually any type of ready-to-eat food.
     Salads are typically served at room temperature or chilled, with notable exceptions such as south
     German potato salad which is served warm.",
          "menuNumber": 2,
          "price": 8.25,
          "revision": 0,
          "title": "Salad",
        },
      ],
      "orderCreated": "Mon Jul 02 2019 23:57:40 GMT-060",
      "stage": 2,
    },
  },
}
`;

exports[`Order reducer Adds item to current order with a count of 1 when currentOrder
      is empty and calling ADD_TO_CURRENT_ORDER 1`] = `
Object {
  "currentOrder": Object {
    "0": Object {
      "count": 1,
      "item": Object {
        "__v": 0,
        "_id": "0",
        "archived": false,
        "created": "Now",
        "description": "A hamburger (short: burger) is a sandwich consisting of one or more cooked
    patties of ground meat, usually beef, placed inside a sliced bread roll or bun. The
    patty may be pan fried, grilled, or flame broiled. Hamburgers are often served with
     cheese, lettuce, tomato, onion, pickles, bacon, or chiles; condiments such as ketchup,
     mayonnaise, mustard, relish, or 'special sauce'; and are frequently placed on sesame
    seed buns. A hamburger topped with cheese is called a cheeseburger.",
        "menuNumber": 1,
        "price": 10.5,
        "revision": 0,
        "title": "Hamburger",
      },
    },
  },
  "items": Object {},
}
`;

exports[`Order reducer Adds to the items count when calling ADD_TO_CURRENT_ORDER and the
      item is already in the order 1`] = `
Object {
  "currentOrder": Object {
    "0": Object {
      "count": 2,
      "item": Object {
        "__v": 0,
        "_id": "0",
        "archived": false,
        "created": "Now",
        "description": "A hamburger (short: burger) is a sandwich consisting of one or more cooked
    patties of ground meat, usually beef, placed inside a sliced bread roll or bun. The
    patty may be pan fried, grilled, or flame broiled. Hamburgers are often served with
     cheese, lettuce, tomato, onion, pickles, bacon, or chiles; condiments such as ketchup,
     mayonnaise, mustard, relish, or 'special sauce'; and are frequently placed on sesame
    seed buns. A hamburger topped with cheese is called a cheeseburger.",
        "menuNumber": 1,
        "price": 10.5,
        "revision": 0,
        "title": "Hamburger",
      },
    },
  },
  "items": undefined,
}
`;

exports[`Order reducer Initial state is empty 1`] = `
Object {
  "currentOrder": Object {},
  "items": Object {},
}
`;

exports[`Order reducer Populates orders when calling FETCH_ORDERS 1`] = `
Object {
  "currentOrder": undefined,
  "items": Object {
    "0": Object {
      "__v": 0,
      "_id": "0",
      "address": "232 Blue St.",
      "customerName": "Joe Green",
      "delivery": true,
      "items": Array [
        Object {
          "__v": 0,
          "_id": "0",
          "archived": false,
          "created": "Now",
          "description": "A hamburger (short: burger) is a sandwich consisting of one or more cooked
    patties of ground meat, usually beef, placed inside a sliced bread roll or bun. The
    patty may be pan fried, grilled, or flame broiled. Hamburgers are often served with
     cheese, lettuce, tomato, onion, pickles, bacon, or chiles; condiments such as ketchup,
     mayonnaise, mustard, relish, or 'special sauce'; and are frequently placed on sesame
    seed buns. A hamburger topped with cheese is called a cheeseburger.",
          "menuNumber": 1,
          "price": 10.5,
          "revision": 0,
          "title": "Hamburger",
        },
        Object {
          "__v": 0,
          "_id": "1",
          "archived": false,
          "created": "Now",
          "description": "A salad is a dish consisting of a mixture of small pieces of food, usually vegetables.
    [1][2] However, different varieties of salad may contain virtually any type of ready-to-eat food.
     Salads are typically served at room temperature or chilled, with notable exceptions such as south
     German potato salad which is served warm.",
          "menuNumber": 2,
          "price": 8.25,
          "revision": 0,
          "title": "Salad",
        },
      ],
      "orderCreated": "Mon Jul 01 2019 23:57:40 GMT-060",
      "stage": 0,
    },
    "1": Object {
      "__v": 0,
      "_id": "1",
      "address": "322 Park St.",
      "customerName": "John Anderson",
      "delivery": false,
      "items": Array [
        Object {
          "__v": 0,
          "_id": "1",
          "archived": false,
          "created": "Now",
          "description": "A salad is a dish consisting of a mixture of small pieces of food, usually vegetables.
    [1][2] However, different varieties of salad may contain virtually any type of ready-to-eat food.
     Salads are typically served at room temperature or chilled, with notable exceptions such as south
     German potato salad which is served warm.",
          "menuNumber": 2,
          "price": 8.25,
          "revision": 0,
          "title": "Salad",
        },
      ],
      "orderCreated": "Mon Jul 02 2019 23:57:40 GMT-060",
      "stage": 2,
    },
  },
}
`;

exports[`Order reducer REMOVE_FROM_CURRENT_ORDER item count when
      removing an item from order and count is > 1 1`] = `
Object {
  "currentOrder": Object {
    "0": Object {
      "count": 1,
      "item": Object {
        "__v": 0,
        "_id": "0",
        "archived": false,
        "created": "Now",
        "description": "A hamburger (short: burger) is a sandwich consisting of one or more cooked
    patties of ground meat, usually beef, placed inside a sliced bread roll or bun. The
    patty may be pan fried, grilled, or flame broiled. Hamburgers are often served with
     cheese, lettuce, tomato, onion, pickles, bacon, or chiles; condiments such as ketchup,
     mayonnaise, mustard, relish, or 'special sauce'; and are frequently placed on sesame
    seed buns. A hamburger topped with cheese is called a cheeseburger.",
        "menuNumber": 1,
        "price": 10.5,
        "revision": 0,
        "title": "Hamburger",
      },
    },
  },
  "items": undefined,
}
`;

exports[`Order reducer REMOVE_FROM_CURRENT_ORDER removes item when
      removing an item from order and count is <= 1 1`] = `
Object {
  "currentOrder": Object {},
  "items": undefined,
}
`;

exports[`Order reducer Returns given state when an action doesn't match 1`] = `
Object {
  "mockState": "MockState",
}
`;

exports[`Order reducer Updates given order in items to be payload when calling UPDATE_ORDER 1`] = `
Object {
  "currentOrder": undefined,
  "items": Object {
    "0": Object {
      "__v": 0,
      "_id": "0",
      "address": "UPDATED INFO",
      "customerName": "Joe Green",
      "delivery": true,
      "items": Array [
        Object {
          "__v": 0,
          "_id": "0",
          "archived": false,
          "created": "Now",
          "description": "A hamburger (short: burger) is a sandwich consisting of one or more cooked
    patties of ground meat, usually beef, placed inside a sliced bread roll or bun. The
    patty may be pan fried, grilled, or flame broiled. Hamburgers are often served with
     cheese, lettuce, tomato, onion, pickles, bacon, or chiles; condiments such as ketchup,
     mayonnaise, mustard, relish, or 'special sauce'; and are frequently placed on sesame
    seed buns. A hamburger topped with cheese is called a cheeseburger.",
          "menuNumber": 1,
          "price": 10.5,
          "revision": 0,
          "title": "Hamburger",
        },
        Object {
          "__v": 0,
          "_id": "1",
          "archived": false,
          "created": "Now",
          "description": "A salad is a dish consisting of a mixture of small pieces of food, usually vegetables.
    [1][2] However, different varieties of salad may contain virtually any type of ready-to-eat food.
     Salads are typically served at room temperature or chilled, with notable exceptions such as south
     German potato salad which is served warm.",
          "menuNumber": 2,
          "price": 8.25,
          "revision": 0,
          "title": "Salad",
        },
      ],
      "orderCreated": "Mon Jul 01 2019 23:57:40 GMT-060",
      "stage": 0,
    },
    "1": Object {
      "__v": 0,
      "_id": "1",
      "address": "322 Park St.",
      "customerName": "John Anderson",
      "delivery": false,
      "items": Array [
        Object {
          "__v": 0,
          "_id": "1",
          "archived": false,
          "created": "Now",
          "description": "A salad is a dish consisting of a mixture of small pieces of food, usually vegetables.
    [1][2] However, different varieties of salad may contain virtually any type of ready-to-eat food.
     Salads are typically served at room temperature or chilled, with notable exceptions such as south
     German potato salad which is served warm.",
          "menuNumber": 2,
          "price": 8.25,
          "revision": 0,
          "title": "Salad",
        },
      ],
      "orderCreated": "Mon Jul 02 2019 23:57:40 GMT-060",
      "stage": 2,
    },
  },
}
`;
